internal/wait/value.go

```diff
@@ -0,0 +1,94 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package wait provides a simple mechanism to wait for a value to be set.
+package wait
+
+import (
+	"context"
+	"sync"
+)
+
+// Value is a simple mechanism to wait for a value to be set. It's like semaphore, but for values.
+//
+//nolint:govet
+type Value[T any] struct {
+	mx    sync.Mutex
+	set   chan struct{}
+	value T
+}
+
+// Set sets the value and unblocks any waiting goroutines.
+func (wv *Value[T]) Set(value T) {
+	wv.mx.Lock()
+	defer wv.mx.Unlock()
+
+	wv.value = value
+
+	if wv.set == nil {
+		wv.set = make(chan struct{})
+	}
+
+	select {
+	case <-wv.set:
```

Comment:
while the mutex is locked, I don't see how this can execute?

Reply:
This only checks if `wv.set` was already closed so it will not close it again. If it was closed, then we return immediately. Essentially this will never block.

Other methods do not hold the mutex for the whole duration of the call so they don't block either.

Reply:
ah... it works if wv.set != nil before we call